diff --git a/include/lala/logic/algorithm.hpp b/include/lala/logic/algorithm.hpp
index d782312..c657b2e 100644
--- a/include/lala/logic/algorithm.hpp
+++ b/include/lala/logic/algorithm.hpp
@@ -382,7 +382,7 @@ CUDA NI std::optional<F> negate(const F& f) {
   return {};
 }
 
-CUDA Sig negate_arithmetic_comparison(Sig sig) {
+inline CUDA Sig negate_arithmetic_comparison(Sig sig) {
   switch(sig) {
     case EQ: return NEQ;
     case NEQ: return EQ;
diff --git a/include/lala/simplifier.hpp b/include/lala/simplifier.hpp
index e11aaa6..4215cf7 100644
--- a/include/lala/simplifier.hpp
+++ b/include/lala/simplifier.hpp
@@ -252,6 +252,22 @@ public:
     }
   }
 
+  template <class Alloc, class Abs, class Env>
+  CUDA auto get_value_of(const LVar<Alloc>& vname, const Env& benv, const Abs& b) const {
+    assert(env.variable_of(vname).has_value());
+    const auto& local_var = env.variable_of(vname)->get();
+    assert(local_var.avar_of(store_aty).has_value());
+    int rep = equivalence_classes[local_var.avar_of(store_aty)->vid()];
+    const auto& rep_name = env.name_of(AVar{store_aty, rep});
+    auto benv_variable = benv.variable_of(rep_name);
+    if(benv_variable.has_value()) {
+      return b.project(benv_variable->get().avars[0]).lb();
+    }
+    else {
+      return constants[rep].lb();
+    }
+  }
+
 private:
   /** \return `true` if mask[i] was changed. */
   CUDA local::B eliminate(battery::dynamic_bitset<memory_type, allocator_type>& mask, size_t i) {
